### PR TITLE
node_cache: add a way to undo unlinks and moves

### PR DESCRIPTION
**Pull request by strib**
Issue: KBFS-2076

### CODE REVIEW

File: libkbfs/interfaces.go [hunk 1]
```diff
@@ -1683,17 +1683,20 @@ type NodeCache interface {
...
+	// found, it returns an error.
+	Move(ref BlockRef, newParent Node, newName string) (func(), error)
 	// Unlink set the corresponding node's parent to nil and caches
 	// the provided path in case the node is still open. NodeCache
 	// ignores the call when ptr is not cached.  The path is required
 	// because the caller may have made changes to the parent nodes
 	// already that shouldn't be reflected in the cached path.
-	// Returns whether a node was actually updated.
-	Unlink(ref BlockRef, oldPath path, oldDe DirEntry) bool
+	// Returns whether a node was actually updated.  It returns a
```
Comment (jzila) [inline libkbfs/interfaces.go]: No longer true, it only returns a function.

File: libkbfs/node_cache.go [hunk 2]
```diff
@@ -193,31 +193,38 @@ func (ncs *nodeCacheStandard) Move(
...
+	oldName := entry.core.pathNode.Name
+
 	entry.core.parent = newParentNS
 	entry.core.pathNode.Name = newName
-	return nil
+
+	return func() {
+		entry.core.parent = oldParent
+		entry.core.pathNode.Name = oldName
+	}, nil
```
Comment (jzila) [inline libkbfs/node_cache.go]: This is a pretty clever approach. Nice!

File: libkbfs/node_cache_test.go [hunk 3]
```diff
@@ -229,6 +229,12 @@ func TestNodeCacheMoveSuccess(t *testing.T) {
 	if ncs.nodes[childPtr2.Ref()].refCount != 1 {
 		t.Errorf("Child1 has wrong refcount: %d", ncs.nodes[childPtr2.Ref()].refCount)
 	}
+
+	undoMove()
+	if childNode2.GetBasename() != "child2" {
```
Comment (jzila) [inline libkbfs/node_cache_test.go]: I guess it's fine not to check the parent?

Comment (strib) [inline libkbfs/node_cache_test.go]: I was just being lazy.  Added a check.

File: libkbfs/interfaces.go [hunk 4]
```diff
@@ -1683,17 +1683,21 @@ type NodeCache interface {
...
-	// If newParent is nil, it treats the ptr's corresponding node as
-	// being unlinked from the old parent completely.
-	Move(ref BlockRef, newParent Node, newName string) error
+	// is not cached.  If newParent is nil, it treats the ptr's
+	// corresponding node as being unlinked from the old parent
+	// completely. If successful, it returns a function that can be
+	// called to undo the effect of the move (or `nil` if nothing
+	// needs to be done); if newParent cannot be found, it returns an
+	// error.
+	Move(ref BlockRef, newParent Node, newName string) (func(), error)
```
Comment (songgao) [inline libkbfs/interfaces.go]: Would you mind write out the name of returned variable e.g. `(undo func(), err error)`? It'd make autocompletion plugins show the meaning of them. Same for `Unlink` below, and their implementations if it's not too much to change :)

Comment (strib) [inline libkbfs/interfaces.go]: @songgao: sure.  Everywhere, or just here in the interface definition?

Comment (songgao) [inline libkbfs/interfaces.go]: Everywhere would be great! The reason is sometimes it's called using the actual type, and the completer would pull the concrete type's method signature

**Pull request merged**